### PR TITLE
vk/dx12: Enable/fix separate back and front lighting

### PR DESCRIPTION
**Pull request by kd-11**
Implement separate front and back diffuse lighting for vulkan.
Also fixes 2-sided diffuse color for dx12 if the vertex shader does not write both outputs separately.
Cosmetic dx12 fix by hiding blended color-key blending if blending is done using a straight forward (factor, 1-factor) equation. A real implementation is still pending.

~~NOTE:
While implementing 2 sided diffuse and specular for vulkan, I noticed that enabling this causes bugs similar to those seen in dx12 in outrun online arcade. For some scenes, 2 sided lighting is enabled and back_diffuse_color is output by the vertex shader. However using the front face test to select correct color gives incorrect results in some cases. Note however, that this is not constant and some other scenes seem to expect the whole flag to be ignored (ingame HUD and skybox).
Disabling the diffuse lighting on dx12 fixes lighting there as well.~~

### CODE REVIEW

**Comment by kd-11** [comment]
Hmm. I think I've figured this out.

File: rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp [hunk 1]
```diff
@@ -202,6 +230,51 @@ void VKFragmentDecompilerThread::insertMainStart(std::stringstream & OS)
...
+							OS << "	vec4 spec_color = back_spec_color;\n";
+						}
+						else
+							OS << "	vec4 spec_color = front_spec_color;\n";
+					}
+
+					continue;
+				}
+				
+				/*
```
Comment (paulsapps) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: Shouldn't all this just be deleted?

File: rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp [hunk 2]
```diff
@@ -202,6 +230,51 @@ void VKFragmentDecompilerThread::insertMainStart(std::stringstream & OS)
...
+				//Disable separable diffuse color until we know more about the control flags
+				//Enabling this will break 2 sided lighting in outrun online
+
+				else if (PI.name == "diff_color")
```
Comment (paulsapps) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: Shouldn't all this commented out stuff just be deleted?

**Comment by kd-11** [comment]
@raven02 Needs nvidia testing. Currently this only affects vulkan. I've figured out the problems with d3d12 as well and will add to this pr before it is ready for merging.

**Comment by raven02** [comment]
@kd-11 .Will test it out shortly

**Comment by raven02** [comment]
Just tested . Outrun looks good on Vulkan

**Comment by kd-11** [comment]
Updated with dx12 fix. Once it is confirmed to work I can clean up the commits.

**Comment by raven02** [comment]
DX12 looks good as well and inline with Vulkan .

untitled

**Comment by vlj** [comment]
thanks for the fix !

**Comment by kd-11** [comment]
Something about this changeset hangs 1942: Joint Strike. Do not merge

**Comment by Nekotekina** [comment]
This PR breaks graphics in Hatsune Miku: Project Diva F

**Comment by Nekotekina** [comment]
bug1
bug2

**Comment by kd-11** [comment]
seems to be a front face culling issue. culling was disabled on vulkan before this pr

**Pull request merged**